Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h

```diff
@@ -0,0 +1,149 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if USE(GBM)
+#include "DMABufBuffer.h"
+#include "GLDisplay.h"
+#include "IntSize.h"
+#include <wtf/OptionSet.h>
+#include <wtf/unix/UnixFileDescriptor.h>
+
+struct gbm_bo;
+struct gbm_device;
+typedef void* EGLImage;
+typedef intptr_t EGLAttrib;
+
+namespace WebCore {
+
+class IntRect;
+
+// Use MemoryMappedGPUBuffer to create a OpenGL texture, that's baked by a dma-buf.
+class MemoryMappedGPUBuffer {
+    WTF_MAKE_NONCOPYABLE(MemoryMappedGPUBuffer);
+    WTF_MAKE_FAST_ALLOCATED();
+public:
+    ~MemoryMappedGPUBuffer();
+
+    enum class BufferFlag : uint8_t {
+        ForceLinear = 1 << 0
+    };
+
+    // Will only return a MemoryMappedGPUBuffer, if gbm_bo allocation + mapping to userland + EGLImage creation succeeded.
+    static std::unique_ptr<MemoryMappedGPUBuffer> create(const IntSize&, OptionSet<BufferFlag>);
+
+    const IntSize& size() const { return m_size; }
+    const OptionSet<BufferFlag>& flags() const { return m_flags; }
+
+    // Map dma-buf into memory, if not yet mapped.
+    bool mapIfNeeded();
+    void unmapIfNeeded();
+
+    // Export gbm_bo buffer as dma-buf and wrap in EGLImage.
+    EGLImage createEGLImageFromDMABuf();
+
+    // Update the underlying data of the dma-buf, as often as desired.
+    // You need to obtain an AccessScope, fencing the write operation.
+    class AccessScope;
+    void updateContents(AccessScope&, const void* srcData, const IntRect& targetRect, unsigned bytesPerLine);
+    void updateContents(AccessScope&, const MemoryMappedGPUBuffer& srcBuffer, const IntRect& targetRect);
+
+    // You need to obtain an AccessScope, fencing the read operation.
+    std::span<uint32_t> mappedDataSpan(AccessScope&) const;
+
+    class AccessScope {
+        WTF_MAKE_NONCOPYABLE(AccessScope);
+        WTF_MAKE_FAST_ALLOCATED();
+    public:
+        ~AccessScope();
+
+        enum class Mode : bool {
+            Read,
+            Write
+        };
+
+        static std::unique_ptr<AccessScope> create(MemoryMappedGPUBuffer&, Mode);
+
+        const Mode& mode() const { return m_mode; }
+        const MemoryMappedGPUBuffer& buffer() const { return m_buffer; }
+
+    private:
+        AccessScope(MemoryMappedGPUBuffer&, Mode);
+
+        MemoryMappedGPUBuffer& m_buffer;
+        Mode m_mode { Mode::Read };
+    };
+
+    bool isMapped() const { return !!m_mappedData; }
+    bool isLinear() const;
+
+private:
+    MemoryMappedGPUBuffer(const IntSize&, OptionSet<BufferFlag>);
+
+    enum class DMABufSyncFlag : uint8_t {
+        Start = 1 << 0,
+        End   = 1 << 1,
+        Read  = 1 << 2,
+        Write = 1 << 3
+    };
+
+    bool performDMABufSyncSystemCall(OptionSet<DMABufSyncFlag> flags);
+    bool allocate(struct gbm_device*, const GLDisplay::DMABufFormat&);
+    bool createDMABufFromGBMBufferObject();
+    UnixFileDescriptor exportGBMBufferObjectAsDMABuf(unsigned planeIndex);
+
+    bool isLocked() const { return m_isLocked; }
```

Comment:
Do we really need a private getter for this private member?

Reply:
Removed.